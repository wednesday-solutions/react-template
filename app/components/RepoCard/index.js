/**
 *
 * RepoCard
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { Card } from '@mui/material';
import T from '@components/T';
import If from '@components/If';
import isEmpty from 'lodash/isEmpty';

const CustomCard = styled(Card)`
  && {
    margin: 1rem 0;
    padding: 1rem;
  }
`;

export function RepoCard({ name, fullName, stargazersCount }) {
  return (
    <CustomCard data-testid="repo-card">
      <If condition={!isEmpty(name)} otherwise={<T data-testid="name-unavailable" id="repo_name_unavailable" />}>
        <T data-testid="name" id="repository_name" values={{ name: name }} />
      </If>
      <If
        condition={!isEmpty(fullName)}
        otherwise={<T data-testid="fullName-unavailable" id="repo_full_name_unavailable" />}
      >
        <T data-testid="fullName" id="repository_full_name" values={{ fullName: fullName }} />
      </If>
      <If condition={stargazersCount} otherwise={<T data-testid="stargazers-unavaiable" id="repo_stars_unavailable" />}>
        <T data-testid="stargazers" id="repository_stars" values={{ stars: stargazersCount }} />
      </If>
    </CustomCard>
  );
}

RepoCard.propTypes = {
  name: PropTypes.string,
  fullName: PropTypes.string,
  stargazersCount: PropTypes.number
};

export default RepoCard;
