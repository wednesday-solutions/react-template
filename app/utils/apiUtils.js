import { create } from 'apisauce';
import snakeCase from 'lodash/snakeCase';
import camelCase from 'lodash/camelCase';
import { mapKeysDeep } from './index';

const { GITHUB_URL } = process.env;
const apiClients = {
  github: null,
  default: null
};
export const getApiClient = (type = 'github') => apiClients[type];
export const generateApiClient = (type = 'github') => {
  switch (type) {
    case 'github':
      apiClients[type] = createApiClientWithTransForm(GITHUB_URL);
      return apiClients[type];
    default:
      apiClients.default = createApiClientWithTransForm(GITHUB_URL);
      return apiClients.default;
  }
};

export const createApiClientWithTransForm = baseURL => {
  const api = create({
    baseURL,
    headers: { 'Content-Type': 'application/json' }
  });
  api.addResponseTransform(response => {
    const { ok, data } = response;
    if (ok && data) {
      response.data = mapKeysDeep(data, keys => camelCase(keys));
    }
    return response;
  });

  api.addRequestTransform(request => {
    const { data } = request;
    if (data) {
      request.data = mapKeysDeep(data, keys => snakeCase(keys));
    }
    return request;
  });
  return api;
};
