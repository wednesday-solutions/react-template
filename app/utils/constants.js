export const SCREEN_BREAK_POINTS = {
  mobile: 320,
  tablet: 768,
  desktop: 992
};
