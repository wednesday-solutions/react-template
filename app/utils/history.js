import { createBrowserHistory } from 'history';
import { isUAT } from './index';

export function getBaseName() {
  if (isUAT()) {
    return `/${process.env.BRANCH_NAME}`;
  }

  return '';
}

const history = createBrowserHistory({ basename: getBaseName() });
export default history;
