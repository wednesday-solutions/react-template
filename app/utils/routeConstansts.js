export default {
  repos: {
    route: '/',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  itunes: {
    route: '/itunes',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  }
};
