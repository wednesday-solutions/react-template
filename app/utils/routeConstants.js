export default {
  repos: {
    route: '/',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  }
};
