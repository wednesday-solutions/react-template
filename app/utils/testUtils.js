import React from 'react';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { browserHistory, BrowserRouter, Router } from 'react-router-dom';
import { ThemeProvider } from '@mui/material/styles';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import { IntlGlobalProvider } from '@components/IntlGlobalProvider';
import { theme } from '@containers/App';

export const renderWithIntl = (children) =>
  render(
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <IntlGlobalProvider>{children}</IntlGlobalProvider>
    </IntlProvider>
  );

export const renderProvider = (children, history) => {
  const store = configureStore({}, browserHistory).store;
  return render(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider theme={theme}>
          {history ? <Router history={history}>{children}</Router> : <BrowserRouter>{children}</BrowserRouter>}
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = (ok, data) => ({
  ok,
  data
});

export const getStyles = () => {};
