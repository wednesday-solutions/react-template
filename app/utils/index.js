import find from 'lodash/find';
import get from 'lodash/get';
import { i18n } from '@lingui/core';
import routeConstants from './routeConstants';

/**
 * Get details of the current route from the route config.
 * If location.pathname doesn't match any entry in the routeConstants it will return null
 * @author mac
 * @date 2020-05-07
 * @param {any} location
 * @returns {any}
 */
export const getCurrentRouteDetails = (location) => {
  if (!get(location, 'pathname')) {
    return null;
  }
  const route = find(
    Object.keys(routeConstants),
    (key) => routeConstants[key].route === location.pathname || `${routeConstants[key].route}/` === location.pathname
  );
  if (route) {
    return routeConstants[route];
  }
  return null;
};
export const mapKeysDeep = (obj, fn) => {
  if (Array.isArray(obj)) {
    return obj.map((val) => mapKeysDeep(val, fn));
  }
  if (typeof obj === 'object') {
    return Object.keys(obj).reduce((acc, current) => {
      const key = fn(current);
      const val = obj[current];
      acc[key] = val !== null && typeof val === 'object' ? mapKeysDeep(val, fn) : val;
      return acc;
    }, {});
  }
  return obj;
};

export const isLocal = () => {
  try {
    const local = JSON.parse(process.env.IS_LOCAL);
    return typeof local === 'boolean' && local;
  } catch {
    // continue regardless of error
  }
  return false;
};

export function isUAT() {
  return process.env.ENVIRONMENT_NAME === 'development' && process.env.NODE_ENV === 'production';
}

export function isProd() {
  return process.env.ENVIRONMENT_NAME === 'production' && process.env.NODE_ENV === 'production';
}

export const translate = (id, values = {}) => i18n._({ id, values });
