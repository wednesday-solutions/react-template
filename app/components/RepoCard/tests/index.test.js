/**
 *
 * Tests for RepoCard
 *
 */

import React from 'react';
import { renderWithIntl } from '@utils/testUtils';
import RepoCard from '../index';
import { translate } from '@app/utils/index';

describe('<RepoCard />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<RepoCard />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 RepoCard component', () => {
    const { getAllByTestId } = renderWithIntl(<RepoCard />);
    expect(getAllByTestId('repo-card').length).toBe(1);
  });

  it('should render the repository details inside the card', () => {
    const repoName = 'react-template';
    const fullName = 'wednesday-solutions/react-template';
    const stargazersCount = 200;
    const { getByTestId } = renderWithIntl(
      <RepoCard name={repoName} fullName={fullName} stargazersCount={stargazersCount} />
    );
    expect(getByTestId('name')).toHaveTextContent(repoName);
    expect(getByTestId('fullName')).toHaveTextContent(fullName);
    expect(getByTestId('stargazers')).toHaveTextContent(stargazersCount);
  });

  it('should render the repository unavailable messages in case any props are unavailable or have falsy values', () => {
    const repoUnavailable = translate('repo_name_unavailable');
    const fullNameUnavailable = translate('repo_full_name_unavailable');
    const stargazersUnavailable = translate('repo_stars_unavailable');
    const { getByTestId } = renderWithIntl(<RepoCard />);
    expect(getByTestId('name-unavailable')).toHaveTextContent(repoUnavailable);
    expect(getByTestId('fullName-unavailable')).toHaveTextContent(fullNameUnavailable);
    expect(getByTestId('stargazers-unavaiable')).toHaveTextContent(stargazersUnavailable);
  });
});
