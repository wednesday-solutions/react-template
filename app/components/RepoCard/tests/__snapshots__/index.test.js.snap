// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepoCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root e6pkn6t0 css-16zkfrb-MuiPaper-root-MuiCard-root"
      data-testid="repo-card"
    >
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="name-unavailable"
      >
        Repository name is unavailable
      </p>
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="fullName-unavailable"
      >
        Repository full name unavaiable
      </p>
      <p
        class="css-lh9t18 egtqi0h0"
        data-testid="stargazers-unavaiable"
      >
        Repository stars are unavaiable
      </p>
    </div>
  </div>
</body>
`;
